File: javascript/affine-cipher/affine-cipher.js
const ALPHABET_LEN = 26
const CODE_A = 'a'.charCodeAt(0)

export const encode = (phrase, key) => {
  let a = key.a
  let b = key.b

  if (!isCoPrime(a, ALPHABET_LEN))
    throw new Error('a and m must be coprime.')

  return group(phrase.toLowerCase().split('').map((ch) => encodeChar(ch, a, b)).join(''))
};

const encodeChar = (ch, a, b) =>
{
  if (ch >= 'a' && ch <= 'z')
  {
    let charIndex = ch.charCodeAt(0) - CODE_A
    let e = ((a * charIndex) + b) % ALPHABET_LEN
    return String.fromCharCode(e + CODE_A)
  }
  else if (ch >= '0' && ch <= '9')
      return ch
  return ''
}

const decodeChar = (ch, a, b) =>
{
  if (ch >= 'a' && ch <= 'z')
  {
    let charIndex = ch.charCodeAt(0) - CODE_A
    let e = (mmi(a, ALPHABET_LEN) * (charIndex - b)) % ALPHABET_LEN
    while (e < 0)
      e += ALPHABET_LEN
    return String.fromCharCode(e + CODE_A)
  }
  else if (ch >= '0' && ch <= '9')
      return ch
  return ''
}

const mmi = (a, m) => 
{
  for(let i = 1; i < m; i++)
    if ((a * i) % m === 1)
      return i;
  throw new Error(`No MMI for (${a}, ${m})`)
}


const group = (phrase, batchSize = 5) => {
  let ret = ''
  let first = true
  for(let i = 0; i < phrase.length; i+=batchSize)
  {
    if (first)
      first = false
    else
      ret += ' '
    ret += phrase.substring(i, i + 5)
  }
  return ret
}

const isCoPrime = (a, b) => {
  for(let i = 2; i <= a && i <= b; i++)
    if (a % i === 0 && b % i === 0)
      return false
  return true
}

export const decode = (phrase, key) => {
  let a = key.a
  let b = key.b

  if (!isCoPrime(a, ALPHABET_LEN))
    throw new Error('a and m must be coprime.')

  return phrase.toLowerCase().split('').map((ch) => decodeChar(ch, a, b)).join('')

};

File: python/affine-cipher/affine_cipher.py
"""
Affine Cipher Implementation
Why yes I did make the bulk of the implementation into a large unreadable list-comprehension just because I could.
"""


ALPHABET = 'abcdefghijklmnopqrstuvwxyz'
ALPHABET_LEN = len(ALPHABET)
ALLOWED_CHARS = ALPHABET + '0123456789'

def encode(plain_text, a, b):
    """
    Encode a plain text value with keys a and b
    :param plain_text: The text to encrypt
    :param a: The first encoding key. Must be co-prime to size of encoding alphabet (26 for a-z)
    :param b: The second encoding key. Neither a or b may be a multiple of each other (3 and 6 would be bad because 6=3*2, but 5, 6 is ok)
    :returns: Encrypted text broken into batches of length 5 with spaces inbetween.
    :raises: ValueError if no Modular Multiplicative Inverse is found.
    """# E(x) = (ai + b) mod m
    mmi_value = mmi(a, ALPHABET_LEN)
    return group(''.join([ALPHABET[(a * ALPHABET.find(char) + b) % ALPHABET_LEN] 
                          if char in ALPHABET else char for char in 
                          [char for char in plain_text.lower() if char in ALLOWED_CHARS]]))

def decode(ciphered_text, a, b):
    """
    Decode a ciphered text that was generated with keys a and b
    :param ciphered_text: The encrypted text
    :param a: The first encoding key. Must be co-prime to size of encoding alphabet (26 for a-z)
    :param b: The second encoding key. Neither a or b may be a multiple of each other (3 and 6 would be bad because 6=3*2, but 5, 6 is ok)
    :returns: Decrypted text.
    :raises: ValueError if no Modular Multiplicative Inverse is found.
    """
    # D(y) = (a^-1)(y - b) mod m
    mmi_value = mmi(a, ALPHABET_LEN)
    return ''.join([ALPHABET[(mmi_value*(ALPHABET.find(char) - b)) % ALPHABET_LEN] 
                    if char in ALPHABET else char for char in 
                    [char for char in ciphered_text.lower() if char in ALLOWED_CHARS]])

def group(text, size = 5):
    """Return a string broken up into chunks of size with spaces inbetween
    :param text: the string to break into chunks
    :param size: the desired size of the chunks (default is 5)
    :returns: copy of the input string with a space inserted after every size characters
    """
    return ' '.join([(text[i:i+size]) for i in range(0, len(text), size)])

def mmi(a, m):
    """
    Find the Modular Multiplicative Inverse of A and M
    :param a: One of the keys
    :param m: The size of the key alphabet
    :returns: Value X such that a*x % m == 1 
    :raises: Value error if no match found.
    """
    for i in range(1, m + 1):
        if (a * i) % m == 1:
            return i
    raise ValueError('a and m must be coprime.')
